data/registry_model/registry_proxy_model.py

```diff
@@ -131,6 +130,51 @@ def lookup_repository(
             state=repo.state if repo is not None else None,
         )
 
+    def _check_blob_upload_and_prune(
+        self, repo_ref: RepositoryReference, upstream_manifest: ManifestInterface
+    ) -> None:
+        """
+        Checks whether the given manifest fits within the quota size for the namespace
+        the repository is part of. If it doesn't, it prunes older tags in the namespace
+        by marking them expired which is eventually garbage collected by the gc worker.
+
+        Raises QuotaExceededException if the given tag is larger than the max quota
+        allotted for the namespace or if there are not enough tags to prune to free up space.
+        """
+        curr_ns_size = namespacequota.get_namespace_size(repo_ref.namespace_name)
+
+        # if quota limit is not set for the namespace, skip auto pruning of images
+        quotas = namespacequota.get_namespace_quota_list(repo_ref.namespace_name)
+        if quotas:
+            # currently only one quota per namespace is supported
+            max_ns_quota = quotas[0].limit_bytes
+        else:
+            logger.info("No quota configured")
+            return
+
+        # if given manifest_ref is a manifestList, we skip checking quota limit
+        if upstream_manifest.is_manifest_list:
+            return
+
+        upstream_blob_size = upstream_manifest.layers_compressed_size
+        if upstream_blob_size > max_ns_quota:
+            raise QuotaExceededException
+
+        # check if upstream image fits within the current namespace size
+        if (curr_ns_size + upstream_blob_size) <= max_ns_quota:
+            return
+
+        # fetch tags with least lifetime_end_ms and mark them expired
+        tags = oci.tag.get_tag_with_least_lifetime_end_for_ns(repo_ref.namespace_name)
```

Comment:
You can save a db lookup if you give `repo_ref.id` to `get_tag_with_least_lifetime_end_for_ns` instead of `namespace_name`

Reply:
`repo_ref.id` and `namespace_name` actually refer to different tables/objects - thx @harishsurf for pointing that out.